### PR TITLE
ENH: Add indexing syntax to GroupBy.nth()

### LINKED ISSUE #44589: ENH: add indexing syntax to GroupBy.nth()
####  Proposed Enhancement
#42947 added slice function arguments to groupby.nth(), but not Python slice index expressions.
I would like to be able to use Python slice notation as an alternative to the current function call.
e.g.
- groupby.nth[b:e:s] to be the same as groupby.nth(slice(b, e, s)).
- groupby.nth[a, b, c, ...] to be the same as groupby.nth([a, b, c, ...]).

#### API breaking implications

None. This is just syntactic sugar using existing functionality, and the existing API will not be changed.
####  Current functionality
```python
df = df = pd.DataFrame({'A': [1, 1, 2, 1, 2],  'B': [1, 2, 3, 4, 5]}, columns=['A', 'B'])
g = df.groupby("A")
g.nth(slice(None, -1))
 	 	B
 	A 	
 	1 	1
 	1 	2
 	2 	3
```
####  Current behavior
```python
g.nth[:-1]
 	---------------------------------------------------------------------------
 	TypeError                                 Traceback (most recent call last)
 	~\AppData\Local\Temp/ipykernel_18684/1996983145.py in <module>
 	----> 1 g.nth[0:-1]

 	TypeError: 'method' object is not subscriptable
```
####  Desired behavior
```python
g.nth[:-1]
 	 	B
 	A 	
 	1 	1
 	1 	2
 	2 	3
```

### PR DESCRIPTION
**Pull request by johnzangwill**
- [x] closes #44589
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Add index notation to nth for lists of integers and slices

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -902,6 +905,15 @@ def __getattr__(self, attr: str):
             f"'{type(self).__name__}' object has no attribute '{attr}'"
         )
 
+    def __getattribute__(self, attr):
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:908]: can you type args & outputs

File: pandas/tests/groupby/test_nth.py [hunk 2]
```diff
@@ -726,6 +726,15 @@ def test_slice(slice_test_df, slice_test_grouped, arg, expected_rows):
     tm.assert_frame_equal(result, expected)
 
 
+def test_nth_indexed(slice_test_df, slice_test_grouped):
```
Comment (jreback) [inline pandas/tests/groupby/test_nth.py:729]: worth testing some more cases here,? e.g. and can compre vs a `slice()` object?

**Review by jreback — CHANGES_REQUESTED** [review]
i like it. some small commments.

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (johnzangwill) [reply pandas/core/groupby/groupby.py]: Typed the arg, but the output has no type

File: pandas/tests/groupby/test_nth.py (hunk 2 above)
Comment (johnzangwill) [reply pandas/tests/groupby/test_nth.py:729]: Added indexed equivalent to test_slices() and call equivalent to test_nth_indexed(). 
The literal index cannot really be parameterized, but I think that these tests are enough to establish equivalence.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @johnzangwill very nice

**Comment by emilmirzayev** [comment]
Hi. Is this already implemented into pandas? I have a use case with groupby slice which is similar to this. Thank you!

**Comment by johnzangwill** [comment]
> Hi. Is this already implemented into pandas? I have a use case with groupby slice which is similar to this. Thank you!

@emilmirzayev This has been merged to master for release in 1.4, which was scheduled for December 31 2021.

**Pull request merged**